File: lib/tools/steiner.dart
import 'dart:math' as math;
import 'package:flutter/material.dart';
import 'package:flutter/scheduler.dart';
import 'package:flutter_math_fork/flutter_math.dart';
import '../scaffold/tool_scaffold.dart';
import '../theory.dart';

/// Steiner-Kette: `n` Kreise tangieren zwei nicht-schneidende Begrenzungskreise.
///
/// Die exzentrische Variante entsteht durch Möbius-Transformation des
/// Einheitskreises auf den symmetrischen Fall — dieselbe Idee wie im
/// Rust-Prototyp.
class SteinerPage extends StatefulWidget {
  const SteinerPage({super.key});

  @override
  State<SteinerPage> createState() => _SteinerPageState();
}

class _SteinerPageState extends State<SteinerPage>
    with SingleTickerProviderStateMixin {
  static const String _theoryAsset = 'assets/theory/steiner.md';

  int _n = 12;
  double _offset = 0.0;
  double _rotation = 0.0;
  bool _animate = true;

  late final Ticker _ticker;
  Duration _lastTick = Duration.zero;

  @override
  void initState() {
    super.initState();
    _ticker = createTicker(_onTick)..start();
  }

  void _onTick(Duration elapsed) {
    if (!_animate) {
      _lastTick = elapsed;
      return;
    }
    final dtMs = (elapsed - _lastTick).inMilliseconds;
    _lastTick = elapsed;
    setState(() {
      _rotation += dtMs * 0.0005;
    });
  }

  @override
  void dispose() {
    _ticker.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return ToolScaffold(
      title: 'Steiner-Kette',
      controls: _buildControls(),
      canvas: _buildCanvas(),
      reference: const ToolReference(tabs: [
        ReferenceTab(
          label: 'Theorie',
          content: TheoryView(assetPath: _theoryAsset),
        ),
      ]),
    );
  }

  Widget _buildControls() {
    return Padding(
      padding: const EdgeInsets.all(16),
      child: ListView(
        children: [
          Text('Anzahl Kreise (n): $_n'),
          Slider(
            value: _n.toDouble(),
            min: 3,
            max: 24,
            divisions: 21,
            label: '$_n',
            onChanged: (v) => setState(() => _n = v.round()),
          ),
          Text('Verschiebung: ${_offset.toStringAsFixed(2)}'),
          Slider(
            value: _offset,
            min: -0.95,
            max: 0.95,
            onChanged: (v) => setState(() => _offset = v),
          ),
          SwitchListTile(
            title: const Text("Rotation (Steiner's Porism)"),
            value: _animate,
            onChanged: (v) => setState(() => _animate = v),
            contentPadding: EdgeInsets.zero,
          ),
          const SizedBox(height: 16),
          Text(
            'Die Möbius-Transformation verzerrt den Raum so, dass die Kreise '
            'sich immer perfekt berühren.',
            style: Theme.of(context).textTheme.bodySmall,
          ),
          const SizedBox(height: 20),
          const Divider(),
          const SizedBox(height: 8),
          Text('Live-Formel:', style: Theme.of(context).textTheme.labelMedium),
          const SizedBox(height: 8),
          _buildFormula(),
        ],
      ),
    );
  }

  Widget _buildFormula() {
    final sinPiN = math.sin(math.pi / _n);
    final rIn = (1 - sinPiN) / (1 + sinPiN);
    final tex =
        r'\begin{aligned}'
        r'r_{\mathrm{in}} &= \dfrac{1 - \sin(\pi/n)}{1 + \sin(\pi/n)} \\[4pt]'
        r'&= \dfrac{1 - \sin(\pi/' '$_n' r')}{1 + \sin(\pi/' '$_n' r')} '
        r'\approx ' '${rIn.toStringAsFixed(4)}'
        r'\end{aligned}';
    return SingleChildScrollView(
      scrollDirection: Axis.horizontal,
      child: Math.tex(
        tex,
        textStyle: const TextStyle(fontSize: 20),
        onErrorFallback: (err) => Text(
          'LaTeX-Fehler: ${err.message}',
          style: const TextStyle(color: Colors.redAccent),
        ),
      ),
    );
  }

  Widget _buildCanvas() {
    return CustomPaint(
      painter: _SteinerPainter(n: _n, offset: _offset, rotation: _rotation),
      child: const SizedBox.expand(),
    );
  }
}

class _SteinerPainter extends CustomPainter {
  _SteinerPainter({
    required this.n,
    required this.offset,
    required this.rotation,
  });

  final int n;
  final double offset;
  final double rotation;

  /// f(z) = (z + a) / (1 + a·z) — Möbius-Automorphismus des Einheitskreises.
  Offset _moebius(double x, double y, Offset center, double radius) {
    final a = offset;
    final denRe = 1.0 + a * x;
    final denIm = a * y;
    final denSq = denRe * denRe + denIm * denIm;
    final resX = ((x + a) * denRe + y * denIm) / denSq;
    final resY = (y * denRe - (x + a) * denIm) / denSq;
    return Offset(center.dx + resX * radius, center.dy - resY * radius);
  }

  /// Zeichnet einen Kreis im Einheitskreis-Koordinatensystem als Polygon
  /// aus 64 möbiustransformierten Punkten.
  void _drawMappedCircle(
    Canvas canvas,
    Offset center,
    double radius,
    double cx,
    double cy,
    double r,
    Paint stroke, {
    Paint? fill,
  }) {
    const steps = 64;
    final path = Path();
    for (var i = 0; i <= steps; i++) {
      final angle = (i / steps) * 2 * math.pi;
      final px = cx + r * math.cos(angle);
      final py = cy + r * math.sin(angle);
      final p = _moebius(px, py, center, radius);
      if (i == 0) {
        path.moveTo(p.dx, p.dy);
      } else {
        path.lineTo(p.dx, p.dy);
      }
    }
    path.close();
    if (fill != null) canvas.drawPath(path, fill);
    canvas.drawPath(path, stroke);
  }

  @override
  void paint(Canvas canvas, Size size) {
    final center = Offset(size.width / 2, size.height / 2);
    final radius = math.min(size.width, size.height) * 0.45;

    final sinPiN = math.sin(math.pi / n);
    final rIn = (1 - sinPiN) / (1 + sinPiN);
    final rChain = (1 - rIn) / 2;
    final rMid = (1 + rIn) / 2;

    final bgStroke = Paint()
      ..style = PaintingStyle.stroke
      ..strokeWidth = 2.0
      ..color = const Color(0xFF888888);
    final innerFill = Paint()
      ..style = PaintingStyle.fill
      ..color = const Color(0xFF202020);
    _drawMappedCircle(canvas, center, radius, 0, 0, 1, bgStroke);
    _drawMappedCircle(canvas, center, radius, 0, 0, rIn, bgStroke,
        fill: innerFill);

    final chainStroke = Paint()
      ..style = PaintingStyle.stroke
      ..strokeWidth = 1.5
      ..color = const Color(0xFFFFD479);
    final chainFill = Paint()
      ..style = PaintingStyle.fill
      ..color = const Color(0x33000000);
    for (var i = 0; i < n; i++) {
      final theta = rotation + (i / n) * 2 * math.pi;
      final cx = rMid * math.cos(theta);
      final cy = rMid * math.sin(theta);
      _drawMappedCircle(canvas, center, radius, cx, cy, rChain, chainStroke,
          fill: chainFill);
    }
  }

  @override
  bool shouldRepaint(covariant _SteinerPainter old) {
    return old.n != n || old.offset != offset || old.rotation != rotation;
  }
}

File: src/tools/steiner.rs
//! Steiner-Kette: `n` Kreise, die zwei konzentrische Kreise tangieren.
//!
//! Die Kette wird zunächst symmetrisch gerechnet (innerer Kreis im Ursprung)
//! und dann durch eine Möbius-Transformation des Einheitskreises
//! `f(z) = (z + a) / (1 + a·z)` exzentrisch verzerrt. Da Möbius-Abbildungen
//! Kreise und Tangentialitäten erhalten, bleibt die Kette geschlossen —
//! das ist Steiner's Porism.

use crate::formula::Formula;
use crate::tool::Tool;
use eframe::egui;
use std::f32::consts::{PI, TAU};
use std::path::{Path, PathBuf};
use typst::foundations::{Dict, IntoValue};

static FORMULA_TEMPLATE: &str = r##"#import sys: inputs
#set page(width: auto, height: auto, margin: 6pt, fill: rgb("#1e1e1e"))
#set text(font: "Latin Modern Roman", fill: rgb("#f0f0f0"), size: 16pt)
#show math.equation: set text(font: "Latin Modern Math")
#let n = int(inputs.n)
#let s = calc.sin(calc.pi / n)
#let r_in = (1 - s) / (1 + s)
$ r_"in" = (1 - sin(pi/n)) / (1 + sin(pi/n)) = (1 - sin(pi/#n)) / (1 + sin(pi/#n)) approx #calc.round(r_in, digits: 4) $
"##;

pub struct Steiner {
    n: usize,
    offset: f32,
    rotation: f32,
    animate: bool,
    formula: Formula<usize>,
}

impl Steiner {
    pub fn new() -> Self {
        Self {
            n: 12,
            offset: 0.0,
            rotation: 0.0,
            animate: true,
            formula: Formula::new(FORMULA_TEMPLATE, "steiner-formula"),
        }
    }
}

impl Tool for Steiner {
    fn name(&self) -> &str {
        "Steiner-Kette"
    }

    fn controls_ui(&mut self, ui: &mut egui::Ui) {
        ui.add(egui::Slider::new(&mut self.n, 3..=24).text("Anzahl Kreise (n)"));
        ui.add(egui::Slider::new(&mut self.offset, -0.95..=0.95).text("Verschiebung"));
        ui.checkbox(&mut self.animate, "Rotation (Steiner's Porism)");

        ui.add_space(20.0);
        ui.label(
            "Die Möbius-Transformation verzerrt den Raum so, dass die Kreise sich \
             immer perfekt berühren.",
        );

        ui.add_space(16.0);
        ui.separator();
        ui.label("Live-Formel:");
        let mut inputs = Dict::new();
        inputs.insert("n".into(), (self.n as i64).into_value());
        self.formula.show(ui, self.n, inputs);
    }

    fn animate(&mut self, ctx: &egui::Context) {
        if self.animate {
            self.rotation += 0.005;
            ctx.request_repaint();
        }
    }

    fn draw(&mut self, painter: &egui::Painter, rect: egui::Rect) {
        let center = rect.center();
        let radius = rect.width().min(rect.height()) * 0.45;

        // Möbius-Automorphismus des Einheitskreises: f(z) = (z + a) / (1 + a·z).
        let moebius = |x: f32, y: f32| -> egui::Pos2 {
            let a = self.offset;
            let den_real = 1.0 + a * x;
            let den_imag = a * y;
            let den_sq = den_real * den_real + den_imag * den_imag;
            let res_x = ((x + a) * den_real + y * den_imag) / den_sq;
            let res_y = (y * den_real - (x + a) * den_imag) / den_sq;
            egui::pos2(center.x + res_x * radius, center.y - res_y * radius)
        };

        // Kreise werden als Polygone aus 64 möbiustransformierten Punkten gezeichnet.
        let draw_mapped_circle =
            |cx: f32, cy: f32, r: f32, stroke: egui::Stroke, fill: egui::Color32| {
                let points = 64;
                let mut shape = Vec::with_capacity(points);
                for i in 0..points {
                    let angle = (i as f32) / (points as f32) * TAU;
                    let px = cx + r * angle.cos();
                    let py = cy + r * angle.sin();
                    shape.push(moebius(px, py));
                }
                shape.push(shape[0]);
                painter.add(egui::Shape::convex_polygon(shape, fill, stroke));
            };

        let sin_pi_n = (PI / self.n as f32).sin();
        let r_in = (1.0 - sin_pi_n) / (1.0 + sin_pi_n);
        let r_chain = (1.0 - r_in) / 2.0;
        let r_mid = (1.0 + r_in) / 2.0;

        let stroke_bg = egui::Stroke::new(2.0, egui::Color32::from_gray(100));
        draw_mapped_circle(0.0, 0.0, 1.0, stroke_bg, egui::Color32::TRANSPARENT);
        draw_mapped_circle(0.0, 0.0, r_in, stroke_bg, egui::Color32::from_gray(40));

        let stroke_chain = egui::Stroke::new(1.5, egui::Color32::GOLD);
        for i in 0..self.n {
            let theta = self.rotation + (i as f32) / (self.n as f32) * TAU;
            let cx = r_mid * theta.cos();
            let cy = r_mid * theta.sin();
            draw_mapped_circle(cx, cy, r_chain, stroke_chain, egui::Color32::from_black_alpha(50));
        }
    }

    fn theory_source(&self) -> Option<PathBuf> {
        let p = Path::new("docs/build/steiner.typ");
        if p.exists() { Some(p.to_path_buf()) } else { None }
    }
}
